promql/engine.go

```diff
@@ -704,17 +734,23 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	}
 	enh := &EvalNodeHelper{out: make(Vector, 0, biggestLen)}
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
+	tempNumSamples = ev.currentSamples
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
+		ev.currentSamples = tempNumSamples
 		// Gather input vectors for this timestamp.
 		for i := range exprs {
 			vectors[i] = vectors[i][:0]
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
-					if point.T == ts {
+					if point.T == ts && ev.currentSamples < ev.maxSamples {
 						vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
+					} else if ev.currentSamples >= ev.maxSamples {
```

Comment:
The else here and duplicated inverse condition is a bit odd, it seems like it might be accidentally broken bu a future change. I'd move this inside the above if as a sub if.